expression/builtin_op.go

```diff
@@ -426,28 +432,58 @@ func (c *isTrueOrFalseFunctionClass) getFunction(ctx sessionctx.Context, args []
 	case opcode.IsTruth:
 		switch argTp {
 		case types.ETReal:
-			sig = &builtinRealIsTrueSig{bf, c.keepNull}
-			sig.setPbCode(tipb.ScalarFuncSig_RealIsTrue)
+			if c.keepNull {
+				sig = &builtinRealIsTrueWithNullSig{bf}
+				sig.setPbCode(tipb.ScalarFuncSig_RealIsTrueWithNull)
+			} else {
+				sig = &builtinRealIsTrueSig{bf}
+				sig.setPbCode(tipb.ScalarFuncSig_RealIsTrue)
+			}
 		case types.ETDecimal:
-			sig = &builtinDecimalIsTrueSig{bf, c.keepNull}
-			sig.setPbCode(tipb.ScalarFuncSig_DecimalIsTrue)
+			if c.keepNull {
+				sig = &builtinDecimalIsTrueWithNullSig{bf}
+				sig.setPbCode(tipb.ScalarFuncSig_DecimalIsTrueWithNull)
+			} else {
+				sig = &builtinDecimalIsTrueSig{bf}
+				sig.setPbCode(tipb.ScalarFuncSig_DecimalIsTrue)
+			}
 		case types.ETInt:
-			sig = &builtinIntIsTrueSig{bf, c.keepNull}
-			sig.setPbCode(tipb.ScalarFuncSig_IntIsTrue)
+			if c.keepNull {
+				sig = &builtinIntIsTrueWithNullSig{bf}
+				sig.setPbCode(tipb.ScalarFuncSig_IntIsTrueWithNull)
+			} else {
+				sig = &builtinIntIsTrueSig{bf}
+				sig.setPbCode(tipb.ScalarFuncSig_IntIsTrue)
+			}
 		default:
 			return nil, errors.Errorf("unexpected types.EvalType %v", argTp)
 		}
 	case opcode.IsFalsity:
 		switch argTp {
 		case types.ETReal:
-			sig = &builtinRealIsFalseSig{bf, c.keepNull}
-			sig.setPbCode(tipb.ScalarFuncSig_RealIsFalse)
+			if c.keepNull {
+				sig = &builtinRealIsFalseWithNullSig{bf}
```

Comment:
I think you don't need to split multiple sigs actually. You only need to split `setPbCode`.

Reply:
Good catch! Update, PTAL. @breeswish